src/ixdat/techniques/ec_ms.py

```diff
@@ -426,6 +431,231 @@ def siq_ecms_calibration_curve(
         else:
             return cal
 
+    # DECONVOLUTION METHODS BELOW
+    def grab_deconvoluted_current(
+        self, mol, impulse_response, tspan=None, tspan_bg=None, snr=10
+    ):
+        """Return the deconvoluted partial current for a given signal using the
+        algorithm developed by Krempl et al.
+        https://pubs.acs.org/doi/abs/10.1021/acs.analchem.1c00110
+
+        Note, this actually doesnt need the EC data - it is calculated
+        from calibrated MS data. It is only meaningful for ECMS measurements
+        at the moment, justifying placement here.
+
+        Args:
+            mol (str): Name of molecule for which deconvolution is to
+                be carried out.
+            impulse_response (ECMSImpulseResponse): The impulse response must contain all
+                the attribute necessary to calculate a new ECMSImpulseResponse with
+                adjusted time and sample frequency. Therefore currently only CALCULATED
+                ECMSImpulseResponse objects are allowed.
+            tspan (list): Timespan for which the partial current is returned. Needs to
+                contain time zero (for alignment with the model).
+            tspan_bg (list): Timespan that corresponds to the background signal.
+            snr (int): signal-to-noise ratio used for Wiener deconvolution.
+        """
+        # grab the calibrated data
+        t_sig, v_sig = self.grab_flux(mol, tspan=tspan, tspan_bg=tspan_bg)
+
+        # first check that the impulse reponse passed is from parameters, because
+        # otherwise this might not work for now. Should make this work in the future!
+        # There is no reason why a measured impulse response shouldnt be just as useful
+        # for deconvolution, but only possible if dt and duration are the same as in
+        # the measurement that is to be deconvoluted.
+        if impulse_response.ir_type != "calculated":
+            raise TechniqueError(
+                "You need to pass an ECMSImpulseResponse object calculated"
+                "from parameters to calculate deconvoluted currents. "
+                "(for now)"
+            )  # TODO: check instead whether the object contains all the necessary
+            # attributes
+
+        # Check if the one passed already has the correct dt and duration
+        dt = t_sig[1] - t_sig[0]
+        duration = t_sig[-1] - t_sig[0]
+
+        if dt == impulse_response.dt and duration == impulse_response.duration:
+            signal_response = (
+                impulse_response  # no need to recalculate if these parameters fit
+            )
+        else:
+            # re-calculate the impulse response
```

Comment:
I think this will have to disappear.

Reply:
What do you mean?

Reply:
I think I hadn't understood it.
I guess it can't disappear, because you do need a `kernel` and `t_kernel` that match the timespan requested. But if you can generate a new `ECMSImpulseResponse` object using only the info in the previous one and the desired time vector, then `kernel` and `t_kernel` are not the defining characteristics of an `ECMSImpulseResponse`. There should be a method `ECMSImpulseResponse.grab_kernel_for_t()` that does this when called from the original object, avoiding the need to make a second one.
I'll do this when merging with [calculators].